### PR TITLE
Partial Shapes and Types, Part 4m: BatchNorm and backprops

### PR DESCRIPTION
**Pull request by aprocter**
Reasons for the `Changes API` tag:

(1) I removed the `BatchNormBase` class, and updated CPU/nvGPU/IntelGPU backends to deal with this.
(2) I removed the restriction on rank for `BatchNormTrainingBackprop`, because this restriction is inconsistent with the forward ops.

### CODE REVIEW

File: src/ngraph/op/batch_norm.cpp [hunk 1]
```diff
@@ -156,63 +114,59 @@ ngraph::op::BatchNormTrainingBackprop::BatchNormTrainingBackprop(
...
 {
-    if (validate_punt_if_dynamic())
-    {
-        return;
-    }
-
-    set_output_size(3);
+    PartialShape input_and_delta_shape{get_input_partial_shape(INPUT)};
 
-    NODE_VALIDATION_ASSERT(this, get_input_shape(INPUT).size() == 4)
```
Comment (aprocter) [inline src/ngraph/op/batch_norm.cpp]: Note that I have removed this restriction. (Does that merit a `Changes API` tag?)

Comment (aprocter) [reply src/ngraph/op/batch_norm.cpp]: Yeah I'm thinking it probably does... sigh.

File: src/ngraph/op/batch_norm.hpp [hunk 2]
```diff
@@ -102,15 +81,26 @@ namespace ngraph
...
+            double get_eps_value() const { return m_epsilon; }
             virtual std::shared_ptr<Node>
                 copy_with_new_args(const NodeVector& new_args) const override;
 
         protected:
             virtual void generate_adjoints(autodiff::Adjoints& adjoints,
                                            const NodeVector& deltas) override;
+
+        private:
+            enum
```
Comment (rkimballn1) [inline src/ngraph/op/batch_norm.hpp]: Use `enum class`

File: src/ngraph/op/batch_norm.hpp [hunk 3]
```diff
@@ -148,6 +139,18 @@ namespace ngraph
             {
                 throw ngraph_error("Invalid operation");
             }
+
+        private:
+            enum
```
Comment (rkimballn1) [inline src/ngraph/op/batch_norm.hpp]: Use `enum class`

File: src/ngraph/op/batch_norm.hpp (hunk 2 above)
Comment (diyessi) [reply src/ngraph/op/batch_norm.hpp]: `constexpr`

**Comment by aprocter** [comment]
@rkimballn1 I opted for symbolic constants for the input indices. Let me know what you think.

File: src/ngraph/runtime/intelgpu/visualize_tree.cpp [hunk 4]
```diff
@@ -123,10 +123,17 @@ void print_node_parameters(ostringstream& writer, const shared_ptr<Node>& node)
...
+            static_pointer_cast<op::BatchNormInference>(node);
+
+        writer << print_table_row_value("EPS", batch_norm->get_eps_value());
+        break;
+    }
     case OP_TYPEID::BatchNormTraining:
     {
-        const shared_ptr<op::BatchNormBase> batch_norm =
-            static_pointer_cast<op::BatchNormBase>(node);
+        const shared_ptr<op::BatchNormTraining> batch_norm =
```
Comment (aprocter) [inline src/ngraph/runtime/intelgpu/visualize_tree.cpp]: Attn: @shssf. `BatchNormBase` is removed in this PR.

File: test/type_prop.cpp [hunk 5]
```diff
@@ -222,158 +222,1276 @@ TEST(type_prop, batchnorm_rank_less_than_2)
...
+
+TEST(type_prop, batchnorm_inference_partial_input_rank_static_dynamic_zero_channels)
+{
+    PartialShape data_batch_shape{
+        Dimension::dynamic(), 0, Dimension::dynamic(), Dimension::dynamic()};
+    PartialShape gamma_shape{PartialShape::dynamic()};
+    PartialShape beta_shape{PartialShape::dynamic()};
+    PartialShape mean_shape{PartialShape::dynamic()};
+    PartialShape variance_shape{PartialShape::dynamic()};
+    double epsilon = 0.001;
```
[the remark this answers is not in the document]
Comment (aprocter) [reply test/type_prop.cpp]: Go away.
